### DISCUSSION TITLE
Steps to build and deploy the service on kubernetes

### DISCUSSION THREAD

**Discussion by pranay-garg-groww** [featurehub-io/featurehub#907] [Q&A]
Hey,
I am able to build the services with the following command : 
`cd backend && mvn -f pom-first.xml install && cd .. && mvn -T4C -DskipTests clean install`
but not able to create the Docker image from all the individual jars. I understand that you use jib plugin for creating docker images but I cannot find the command to run that. 
Can you guyz pls send a step by step commands for setting up a kubernets pipeline for building and deploying the service.

We are evaluating featurehub for a company wide use case at large scale and need some assistance. It would be really helpful if we can get on a call and understand some internal setup of the service. 

Thanks

**Comment by rvowles** [comment]
Hey there! We don't recommend you build the code unless you are intending to contribute functionality. If you would like to use k8s, please feel free to use the helm chart https://github.com/featurehub-io/featurehub-helm, there is space in the helm chart values for you to add you appropriate Cloud Vendor's required customisations to expose the service.  There is also a `kind` local installation guide for you to run it on k8s locally so you can see how it fits together.

If you would like to follow the Architecture - please have a look at our Architecture diagrams in our documentation and feel free to ask questions here.

**Comment by pranay-garg-groww** [comment]
Hey, thanks for the reply.
I actually intend to contribute functionality and play around with some changes. Can you please share the details of how to build the code to a docker image?

**Comment by rvowles** [comment]
Sure! If you are planning on building a docker image just to the local docker engine, there is a script in the backend folder called `build.sh` - if you run that then it builds the appropriate docker images into your local machine. If you have a registry that you have access to and wish to push the images there automatically, then the `release.sh` script in the same folder does this, but you will need to set the environment variable `OVERRIDE_DOCKER_PREFIX` to specify your repository and prefix. Its just `featurehub` by default at the moment and pushes to Dockerhub and our `featurehub` account. But it will build multi-architecture images.

The front end is built using Flutter, but it does use a docker image to do so, which is on Dockerhub so you should be fine doing that.

So the command you listed which builds all of the code into your local repo, and then the `build.sh` or `release.sh` script builds the images. Note, there are no Dockerfiles in any of the front ends because Jib takes care of all of that (user space docker image building).

**Comment by pranay-garg-groww** [comment]
got it, 
while running the file I am getting the following error
`Status: Downloaded newer image for featurehub/flutter_web:1.8
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
total 3984
-rw-r--r--  1 root root 4075520 Dec 23 10:44 admin-frontend-1.1-SNAPSHOT.tar
drwxr-xr-x 10 root root     320 Dec 23 07:59 app_mr_layer
drwxr-xr-x  2 root root      64 Dec 23 10:44 archive-tmp
-rw-r--r--  1 root root    1818 Dec 23 10:44 build.sh
drwxr-xr-x  2 root root      64 Dec 23 10:44 build_web
drwxr-xr-x  5 root root     160 Dec 23 10:44 classes
drwxr-xr-x  3 root root      96 Dec 23 10:44 generated-sources
drwxr-xr-x 10 root root     320 Nov 23 07:57 open_admin_app
flutter home is /opt/flutter
Flutter version is 3.3.2 dart version is 2.18.1
FLUTTER: Cleaning up after last build
Running "flutter pub get" in app_mr_layer...
===== CRASH =====
si_signo=Segmentation fault(11), si_code=1, si_addr=0x10006
version=2.18.1 (stable) (Tue Sep 13 11:42:55 2022 +0200) on "linux_x64"
pid=17, thread=61, isolate_group=main(0x4003061000), isolate=main(0x40030d8000)
isolate_instructions=4001ece380, vm_instructions=4001ece380
  pc 0x0000ffffae28472b fp 0x00000040103ef380 Unknown symbol
  pc 0x0000ffffacde2aa6 fp 0x00000040103ef3f0 Unknown symbol
  pc 0x0000ffffada401bd fp 0x00000040103ef430 Unknown symbol
  pc 0x0000ffffae02337b fp 0x00000040103ef498 Unknown symbol
  pc 0x0000ffffae022c23 fp 0x00000040103ef4e8 Unknown symbol
  pc 0x0000ffffae282a0c fp 0x00000040103ef560 Unknown symbol
  pc 0x000000400204653d fp 0x00000040103ef600 dart::DartEntry::InvokeCode(dart::Code const&, unsigned long, dart::Array const&, dart::Array const&, dart::Thread*)+0x14d
  pc 0x000000400204637c fp 0x00000040103ef660 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)+0x14c
  pc 0x00000040020487ac fp 0x00000040103ef6b0 dart::DartLibraryCalls::HandleMessage(long, dart::Instance const&)+0x14c
  pc 0x000000400206fe20 fp 0x00000040103efc40 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)+0x350
  pc 0x000000400209934d fp 0x00000040103efcb0 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x14d
  pc 0x0000004002099a2f fp 0x00000040103efd00 dart::MessageHandler::TaskCallback()+0x1df
  pc 0x00000040021bc8e8 fp 0x00000040103efd80 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x148
  pc 0x00000040021bcd3d fp 0x00000040103efdb0 dart::ThreadPool::Worker::Main(unsigned long)+0x6d
  pc 0x000000400212f508 fp 0x00000040103efe70 /opt/flutter/bin/cache/dart-sdk/bin/dart+0x212f508
-- End of DumpStackTrace
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted`
in front-end build

**Comment by rvowles** [comment]
Heya - this is presumably because you are running on an Apple Silicon Mac and it doesn't emulate a AMD64 architecture for certain key areas. You will need to build the image for the flutter-web docker image for Apple Silicon yourself unfortunately, we are all on Intel silicon, and our build servers are all Intel as well. The image is in the folder `pipeline/build/flutter_web` - if you are building locally you can just use the script as is - just `docker image rm` the existing image on your machine. There is a script there that will build the requisite image and the debian base is multi-architecture.

**Comment by pranay-garg-groww** [comment]
I've removed the image and when running the make-image.sh I am getting the following error
` => ERROR [ 3/12] RUN apt-get install -y   git   wget   curl   unzip   lcov   lib32stdc++6   libglu1-mesa                                                                                                   0.9s
------
 > [ 3/12] RUN apt-get install -y   git   wget   curl   unzip   lcov   lib32stdc++6   libglu1-mesa:
#6 0.321 Reading package lists...
#6 0.740 Building dependency tree...
#6 0.823 Reading state information...
#6 0.916 E: Unable to locate package lib32stdc++6
#6 0.916 E: Couldn't find any package by regex 'lib32stdc++6'
------
executor failed running [/bin/sh -c apt-get install -y   git   wget   curl   unzip   lcov   lib32stdc++6   libglu1-mesa]: exit code: 100`

**Comment by rvowles** [comment]
Hey again! I'm back from xmas/new years break, did you resolve this?
